527: Replace DefaultFormattingConversionService with Bean

zersetz-end opened the pull request:
[Issue #527](https://github.com/spring-projects/spring-kafka/issues/527)

pivotal-issuemaster commented:
@zersetz-end Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=528)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=528) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@zersetz-end Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=528)!

artembilan commented:
Thank you for contribution, @zersetz-end !

I didn't check the code yet, but that really would be better to have some test-case to confirm the requested functionality or demonstrate how it fixes some expectations.

artembilan commented on spring-kafka/src/test/java/org/springframework/kafka/annotation/EnableKafkaIntegrationTests.java:
```diff
@@ -976,6 +1017,36 @@ public KafkaListenerErrorHandler voidSendToErrorHandler() {
+					fooConverter()
+			));
+			bean.afterPropertiesSet();
+			return bean.getObject();
```
You don't need to call `afterPropertiesSet()` in the `@Bean` definition method.
You don't need to call `FactoryBean.getObject()` in the `@Bean` definition method.
The Framework takes care about all the lifecycle callbacks and autowiring injections.

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaListenerAnnotationBeanPostProcessor.java:
```diff
@@ -664,7 +665,17 @@ private MessageHandlerMethodFactory createDefaultMessageHandlerMethodFactory() {
+			catch (Exception e) {
+				KafkaListenerAnnotationBeanPostProcessor.this.logger.warn(
+						"Failed to get bean for type ConversionService. "
+								+ "Using DefaultFormattingConversionService.", e);
```
I think we don't need this log message at all. At least on the `info` (or even `debug`) level...
And would be better to say:

> No ConversionService bean in the application context. Using DefaultFormattingConversionService as default instance.

zersetz-end commented:
I refactored the way Converter beans are added to the ConversionService after a little chat in spring-boot, because I wasn't really sure where the default ConversionService bean usually comes from. The title/issue content don't really match the implementation now, but I guess this is the better way of registering custom converters.

Not really sure where to document the feature though.

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaListenerAnnotationBeanPostProcessor.java:
```diff
@@ -690,6 +696,26 @@ protected boolean isEmptyPayload(Object payload) {
+
+		private <T> Collection<T> getBeansOfType(Class<T> type) {
+			if (KafkaListenerAnnotationBeanPostProcessor.this.beanFactory instanceof ListableBeanFactory) {
+				return ((ListableBeanFactory) KafkaListenerAnnotationBeanPostProcessor.this.beanFactory).getBeansOfType(type).values();
```
I checked the call stack and all of these `getBeansOfType()` is not good during `BeanPostProcessor.postProcessAfterInitialization()`. That is an "early BeanFactory initialization syndrome".

If we are going this `add*` way, we should consider to deffer their population until `afterSingletonsInstantiated()` phase.

zersetz-end commented:
I added documentation and changed the copyright dates. English isn't my native language, so if you have any remarks about the wording or the content, feel free to provide me an improved version. :-)

artembilan commented on src/reference/asciidoc/kafka.adoc:
```diff
@@ -1271,6 +1271,31 @@ public void listen(List<Foo> foos, @Header(KafkaHeaders.OFFSET) List<Long> offse
 
 Notice that you can still access the batch headers too.
 
+Starting with _versions 1.3.3_ the `org.springframework.core.convert.ConversionService` used by the default
```
I don't think that we are going to back port it down into the `1.3.x`. (`2.1.1` should be good)
There is a workaround like `MessageHandlerMethodFactory` injection into the `KafkaListenerEndpointRegistrar` as you correctly mention below.

There should be `,` after version.

zersetz-end replied:
I'm aware of the workaround, as i'm currently forced to use it. :-)

But I think the beauty of my solution is that unless you have to programatically configure any other properties of the listener handling that aren't confiurable through the `KafkaProperties`, you basically just have to put `@EnableKafka` somewhere and implement `Converter` for your incoming message types. Or maybe I'm misunderstanding the workaround, because this is what I have to do, to achieve the same behaviour:

``` java
@Component
public class KafkaConfig implements KafkaListenerConfigurer {

    @Autowired
    BeanFactory beanFactory;

    @Autowired
    ConversionService conversionService;

    @Override
    public void configureKafkaListeners(KafkaListenerEndpointRegistrar registrar) {
        registrar.setMessageHandlerMethodFactory(createDefaultMessageHandlerMethodFactory());
    }

    private MessageHandlerMethodFactory createDefaultMessageHandlerMethodFactory() {
        DefaultMessageHandlerMethodFactory defaultFactory = new DefaultMessageHandlerMethodFactory();
        defaultFactory.setBeanFactory(beanFactory);

        ConfigurableBeanFactory cbf =
                (beanFactory instanceof ConfigurableBeanFactory ?
                        (ConfigurableBeanFactory) beanFactory : null);

        defaultFactory.setConversionService(conversionService);

        List<HandlerMethodArgumentResolver> argumentResolvers = new ArrayList<>();

        argumentResolvers.add(new HeaderMethodArgumentResolver(conversionService, cbf));
        argumentResolvers.add(new HeadersMethodArgumentResolver());

        final GenericMessageConverter messageConverter = new GenericMessageConverter(conversionService);
        argumentResolvers.add(new MessageMethodArgumentResolver(messageConverter));
        argumentResolvers.add(new PayloadArgumentResolver(messageConverter) {

            @Override
            protected boolean isEmptyPayload(Object payload) {
                return payload == null || payload instanceof KafkaNull;
            }

        });
        defaultFactory.setArgumentResolvers(argumentResolvers);

        defaultFactory.afterPropertiesSet();
        return defaultFactory;
    }

}
```

artembilan replied:
Seems for me your workaround is overhead.
That should be like this:
```
@Override
public void configureKafkaListeners(KafkaListenerEndpointRegistrar registrar) {
        registrar.setMessageHandlerMethodFactory(defaultMessageHandlerMethodFactory());
}

@Bean
public MessageHandlerMethodFactory defaultMessageHandlerMethodFactory() {
    DefaultMessageHandlerMethodFactory defaultFactory = new DefaultMessageHandlerMethodFactory();
   defaultFactory.setConversionService(conversionService);
   return defaultFactory;
```

zersetz-end replied:
Yes, I noticed that too. But this way it lacks the handling of `KafkaNull`. Which shouldn't cause any problems in my case. I'm also currently looking into adding a `Validator` to the `MessageHandlerMethodFactory'. Something you should consider adding in similar fashion some day. :-D

artembilan replied:
Oh! I see what you mean about `KafkaNull`. Hard to read long code on GH 😄 .

Well, for that purpose we have `DefaultMessageHandlerMethodFactory.setCustomArgumentResolvers()`.

artembilan commented on src/reference/asciidoc/kafka.adoc:
```diff
@@ -1271,6 +1271,31 @@ public void listen(List<Foo> foos, @Header(KafkaHeaders.OFFSET) List<Long> offse
 
+Starting with _versions 1.3.3_ the `org.springframework.core.convert.ConversionService` used by the default
+`org.springframework.messaging.handler.annotation.support.MessageHandlerMethodFactory` to reslove parameters for the invocation
+of a listener method is provided all beans implementing any of the following interfaces:
```
> ... is supplied with all beans ...

artembilan commented on src/reference/asciidoc/kafka.adoc:
```diff
@@ -1271,6 +1271,31 @@ public void listen(List<Foo> foos, @Header(KafkaHeaders.OFFSET) List<Long> offse
+    public Foo convert(String source) {
+        return JAXB.unmarshal(source, Foo.class);
+    }
+}
```
This code sample just doesn't make sense. Please, remove it altogether.
